File: DSA/Algorithms/Recursion/max_element_array.c
// Problem: Find the maximum element in an array using recursion.
// Approach: The max is either the current element, or the max of the
// rest of the array — whichever is larger.
//Base case: last element is trivially the max of a one-element array.

#include<stdio.h>

int findMax(int arr[], int n, int i){
    if(i == n-1){
        return arr[i];
    }
    int maxOfRest = findMax(arr, n, i+1);
    if(arr[i] > maxOfRest)
        return arr[i];
    else
        return maxOfRest;
}

int main(){
    int n;
    printf("enter size of array:");
    scanf("%d", &n);

    int arr[n];
    printf("\nenter array elements:");
    for(int i=0; i<n; i++)
        scanf("%d", &arr[i]);

    int result = findMax(arr, n, 0);
    printf("\nmaximum element is %d", result);

    return 0;
}

File: DSA/Algorithms/Recursion/max_element_array.java
// Problem: Find the maximum element in an array using recursion.
// Approach: Accumulator pattern — carry the "max found so far" forward 
// as a parameter, comparing it against the next element at each step,
// until the last index is reached.

import java.util.Scanner;

public class MaxElement {

    static int findMax(int[] arr, int n, int i, int max){
        if(i == n-1){
            return max;
        }
        if(arr[i+1] > max){
            i++;
            return findMax(arr, n, i, arr[i]);
        }
        else{
            i++;
            return findMax(arr, n, i, max);
        }
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        System.out.print("enter size of array:");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("\nenter array elements:");
        for(int i=0; i<n; i++)
            arr[i] = sc.nextInt();

        int result = findMax(arr, n, 0, arr[0]);
        System.out.print("\nmaximum element is " + result);
    }
}
